history-traversal.md

```diff
@@ -0,0 +1,107 @@
+
+   Note: This is an in-document navigation. Nothing portal-related happens in this case.
+
+1. Let _documentChangesInDelta_ be 0.
+1. Let _historyItems_ be null.
+1. If _isReplacement_, then set _historyItems_ to [the current history item, _targetHistoryItem_].
+1. Otherwise, set _historyItems_ to the history items of this navigable, between the current item, and target item, including the current item and _targetHistoryItem_, in order.
+1. Remove the first item from _historyItems_.
+1. For each _historyItem_ of _historyItems_, if any of the following is true, increment _documentChangesInDelta_:
+   - _historyItem_ has a different browsing context to the previous item in the list.
+   - _historyItem_'s URL is different to the previous item in the list, ignoring the hash portion of the URL in both.
+1. If _documentChangesInDelta_ is greater than 1, then traverse to _targetHistoryItem_ in the regular way and abort these steps.
+
+   Note: To avoid multiple levels of reportaling, reportaling and implicit activation is skipped if the navigation spans across multiple documents
+
+1. Asset: _documentChangesInDelta_ is not 0.
+
+   Note: This should have been catered for in step 1.
+
+1. If _targetHistoryItem_ does not contain a document, and _targetHistoryItem_ has a weak reference to a document that's inside a portal in the current history item's document. then:
+
+   1. Let _document_ be that portal's document.
+   1. For each item in the navigable's session history that has a weak reference to _document_, set its document to _document_, and remove its weak reference to _document_.
+   1. Give the portal a weak reference to the document inside the portal.
+   1. Remove the portal's session history item.
+   1. If adoption is permitted, offer _targetHistoryItem_'s document the opportunity to adopt the current history item. If this opportunity is taken:
+
+      1. A new portal element is created in _targetHistoryItem_'s document.
+      1. Give the portal a weak reference to the current history item's document.
+
+         Note: The rest of the process is picked up in step 12.
+
+1. If _targetHistoryItem_ does not contain a document, then traverse to _targetHistoryItem_ in the regular way and abort these steps.
+1. If _targetHistoryItem_'s document contains a portal with a weak reference to the current history item's document, then:
+   1. Let _documentToPortal_ be the current history item's document.
+   1. Set the portal's history item to a copy of the current history item.
+   1. Remove the portal's weak reference to _documentToPortal_.
+   1. For each item in the navigable's session history, remove its document if the document is _documentToPortal_, and give it a weak reference to _documentToPortal_.
+1. Otherwise, unload the current history item's document.
+
+   Note: This may include excluding it from bfcache.
```

Comment:
I don't quite follow this but I assume there's a reason.

Reply:
I'll improve the language here. During unloading, things can happen that exclude the document from bfcache, such as unload event handlers.